lib/src/RealIpResolver.cc

```diff
@@ -0,0 +1,201 @@
+}
+
+void RealIpResolver::initAndStart(const Json::Value& config)
+{
+    fromHeader_ = config.get("from_header", "x-forwarded-for").asString();
+    attributeKey_ = config.get("attribute_key", "real-ip").asString();
+
+    std::transform(fromHeader_.begin(),
+                   fromHeader_.end(),
+                   fromHeader_.begin(),
+                   [](unsigned char c) { return tolower(c); });
+    if (fromHeader_ == "x-forwarded-for")
+    {
+        useXForwardedFor_ = true;
+    }
+
+    const Json::Value& trustIps = config["trust_ips"];
+    if (!trustIps.isArray())
+    {
+        throw std::runtime_error("Invalid trusted_ips. Should be array.");
+    }
+    for (const auto& elem : trustIps)
+    {
+        std::string ipOrCidr = elem.asString();
+        trustCIDRs_.emplace_back(ipOrCidr);
+    }
+
+    drogon::app().registerPreHandlingAdvice([this](const HttpRequestPtr& req) {
+        const std::string& ipHeader = req->getHeader(fromHeader_);
+        const trantor::InetAddress& peerAddr = req->getPeerAddr();
+        if (ipHeader.empty() || !matchCidr(peerAddr))
+        {
+            // Target header is empty, or
+            // direct peer is already a non-proxy
+            req->attributes()->insert(attributeKey_, peerAddr);
+            return;
+        }
+        // Use a header field which contains a single ip
+        if (!useXForwardedFor_)
+        {
+            trantor::InetAddress addr = parseAddress(ipHeader);
+            if (addr.isUnspecified())
+            {
+                req->attributes()->insert(attributeKey_, peerAddr);
+            }
+            else
+            {
+                req->attributes()->insert(attributeKey_, addr);
+            }
+            return;
+        }
+        // Use x-forwarded-for header, which may contains multiple ip address,
+        // separated by comma
+        XForwardedForParser parser(ipHeader);
+        std::string ip;
+        while (!(ip = parser.getNext()).empty())
+        {
+            trantor::InetAddress addr = parseAddress(ip);
+            if (addr.isUnspecified() || matchCidr(addr))
+            {
+                continue;
+            }
+            req->attributes()->insert(attributeKey_, addr);
+            return;
+        }
+        // No match, use peerAddr
+        req->attributes()->insert(attributeKey_, peerAddr);
+    });
+}
+
+void RealIpResolver::shutdown()
+{
+}
+
+bool RealIpResolver::matchCidr(const trantor::InetAddress& addr) const
+{
+    for (auto& cidr : trustCIDRs_)
+    {
+        if ((addr.ipNetEndian() & cidr.mask_) == cidr.addr_)
+        {
+            return true;
+        }
+    }
+    return false;
+}
+
+RealIpResolver::CIDR::CIDR(const std::string& ipOrCidr)
+{
+    // Find CIDR slash
+    auto pos = ipOrCidr.find('/');
+    std::string ipv4;
+    if (pos != std::string::npos)
+    {
+        // parameter is a CIDR block
+        std::string prefixLen = ipOrCidr.substr(pos + 1);
```

Comment:
can we use `string_view` to avoid string copy?

Reply:
The `CIDR` constructor will only be called in initialization. No need to save here.
And trantor::InetAddress only takes std::string as parameter.